examples/sir_hmc.py

```diff
@@ -0,0 +1,632 @@
+                    dist.ExtendedBinomial(I_prev, prob_i),
+                    obs=I2R)
+        pyro.sample("obs_{}".format(t),
+                    dist.ExtendedBinomial(S2I.clamp(min=0), rho),
+                    obs=datum)
+
+
+# By reparameterizing, we have converted to coordinates to make the model
+# Markov. We have also replaced dynamic integer_interval constraints with
+# easier static integer_interval constraints (although we'll still need good
+# initialization to avoid NANs). Since the discrete latent variables are
+# bounded (by population size), we can enumerate out discrete latent variables
+# and perform HMC inference over the global latents. However enumeration
+# complexity is O(population^4), so this is only feasible for very small
+# populations.
+#
+# Here is an inference approach using an MCMC sampler.
+
+def infer_hmc_enum(args, data):
+    model = reparameterized_discrete_model
+    return _infer_hmc(args, data, model)
+
+
+def _infer_hmc(args, data, model, init_values={}):
+    logging.info("Running inference...")
+    kernel = NUTS(model,
+                  max_tree_depth=args.max_tree_depth,
+                  init_strategy=init_to_value(values=init_values),
+                  jit_compile=args.jit, ignore_jit_warnings=True)
+    energies = []
+
+    def hook_fn(kernel, *unused):
+        e = float(kernel._potential_energy_last)
+        energies.append(e)
+        if args.verbose:
+            logging.info("potential = {:0.6g}".format(e))
+
+    mcmc = MCMC(kernel, hook_fn=hook_fn,
+                num_samples=args.num_samples,
+                warmup_steps=args.warmup_steps)
+    mcmc.run(data, population=args.population)
+    mcmc.summary()
+    if args.plot:
+        import matplotlib.pyplot as plt
+        plt.figure()
+        plt.plot(energies)
+        plt.xlabel("MCMC step")
+        plt.ylabel("potential energy")
+        plt.title("MCMC energy trace")
+        plt.tight_layout()
+
+    samples = mcmc.get_samples()
+    return samples
+
+
+# To scale to large populations, we'll continue to reparameterize, this time
+# replacing each of (S_aux,I_aux) with a combination of a positive real
+# variable and a Bernoulli variable.
+#
+# This is the crux: we can now perform HMC over the real variable and
+# marginalize out the Bernoulli using variable elimination.
+#
+# We first define a helper to create enumerated Bernoulli sites.
+
+SPLINE_ORDER = 1
+
+
+def quantize(name, x_real, min, max):
+    """Randomly quantize in a way that preserves probability mass."""
+    lb = x_real.detach().floor()
+
+    if SPLINE_ORDER == 1:
+        # This linear spline interpolates over the nearest two integers.
+        prob = x_real - lb
+        q = pyro.sample("Q_" + name, dist.Bernoulli(prob))
+        # This cubic spline interpolates over the nearest four integers.
+    elif SPLINE_ORDER == 3:
+        s = x_real - lb
+        ss = s * s
+        t = 1 - s
+        tt = t * t
+        probs = torch.stack([
```

Comment:
i'll need a day or two before i have a chance to check the correctness of the cubic construction

Reply:
https://github.com/fritzo/notebooks/blob/master/cubic_spline_dist.ipynb